File: gateway/src/api-client.ts
/** HTTP client for the Vadgr API. */

const DEFAULT_TIMEOUT = 30_000;
const MAX_RETRIES = 3;
const RETRY_BACKOFF_BASE = 1000;
const ID_PATTERN = /^[a-zA-Z0-9-]+$/;

export class VadgrAPIClient {
  private baseUrl: string;

  constructor(baseUrl = "http://localhost:8000") {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  private validateId(id: string): void {
    if (!ID_PATTERN.test(id)) throw new Error(`Invalid ID: ${id.slice(0, 20)}`);
  }

  async listAgents(): Promise<Record<string, unknown>[]> {
    return this.requestWithRetry(`${this.baseUrl}/api/agents`);
  }

  async runAgent(agentId: string, inputs: Record<string, string>): Promise<Record<string, unknown>> {
    this.validateId(agentId);
    return this.requestWithRetry(`${this.baseUrl}/api/agents/${agentId}/run`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ inputs }),
    });
  }

  async listRuns(): Promise<Record<string, unknown>[]> {
    return this.requestWithRetry(`${this.baseUrl}/api/runs`);
  }

  async getRun(runId: string): Promise<Record<string, unknown>> {
    this.validateId(runId);
    return this.requestWithRetry(`${this.baseUrl}/api/runs/${runId}`);
  }

  async cancelRun(runId: string): Promise<Record<string, unknown>> {
    this.validateId(runId);
    return this.requestWithRetry(`${this.baseUrl}/api/runs/${runId}/cancel`, { method: "POST" });
  }

  async resumeRun(runId: string): Promise<Record<string, unknown>> {
    this.validateId(runId);
    return this.requestWithRetry(`${this.baseUrl}/api/runs/${runId}/resume`, { method: "POST" });
  }

  async getRunLogs(runId: string): Promise<Record<string, unknown>[]> {
    this.validateId(runId);
    return this.requestWithRetry(`${this.baseUrl}/api/runs/${runId}/logs`);
  }

  private async requestWithRetry(url: string, init?: RequestInit): Promise<any> {
    for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
      const resp = await fetch(url, {
        ...init,
        signal: AbortSignal.timeout(DEFAULT_TIMEOUT),
      });
      if (resp.ok) return resp.json();
      if (resp.status >= 500 && attempt < MAX_RETRIES - 1) {
        await new Promise((r) => setTimeout(r, RETRY_BACKOFF_BASE * 2 ** attempt));
        continue;
      }
      throw new Error(`API error ${resp.status}: ${await resp.text()}`);
    }
  }
}

File: gateway/api_client.py
"""HTTP client for talking to the Vadgr API.

Thin wrapper around httpx that the router and notifier use
to list agents, trigger runs, check status, etc.
"""

from __future__ import annotations

import httpx

_TIMEOUT = 30


class VadgrAPIClient:
    """Async client for the Vadgr REST API."""

    def __init__(self, base_url: str = "http://localhost:8000"):
        self._base_url = base_url.rstrip("/")

    async def list_agents(self) -> list[dict]:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.get(f"{self._base_url}/api/agents")
            resp.raise_for_status()
            return resp.json()

    async def get_agent(self, agent_id: str) -> dict:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.get(f"{self._base_url}/api/agents/{agent_id}")
            resp.raise_for_status()
            return resp.json()

    async def run_agent(self, agent_id: str, inputs: dict) -> dict:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.post(
                f"{self._base_url}/api/agents/{agent_id}/run",
                json={"inputs": inputs},
            )
            resp.raise_for_status()
            return resp.json()

    async def list_runs(self, status: str | None = None) -> list[dict]:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            params = {"status": status} if status else {}
            resp = await client.get(f"{self._base_url}/api/runs", params=params)
            resp.raise_for_status()
            return resp.json()

    async def get_run(self, run_id: str) -> dict:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.get(f"{self._base_url}/api/runs/{run_id}")
            resp.raise_for_status()
            return resp.json()

    async def cancel_run(self, run_id: str) -> dict:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.post(f"{self._base_url}/api/runs/{run_id}/cancel")
            resp.raise_for_status()
            return resp.json()

    async def resume_run(self, run_id: str) -> dict:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.post(f"{self._base_url}/api/runs/{run_id}/resume")
            resp.raise_for_status()
            return resp.json()

    async def get_run_logs(self, run_id: str) -> list[dict]:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.get(f"{self._base_url}/api/runs/{run_id}/logs")
            resp.raise_for_status()
            return resp.json()
